ultra/train.py

```diff
@@ -0,0 +1,182 @@
+def train(
+    scenario_info, num_episodes, policy_class, eval_info, timestep_sec, headless, seed
+):
+    torch.set_num_threads(1)
+    total_step = 0
+    finished = False
+
+    # --------------------------------------------------------
+    # Initialize Agent and social_vehicle encoding method
+    # -------------------------------------------------------
+    AGENT_ID = "007"
+
+    spec = make(locator=policy_class)
+    env = gym.make(
+        "ultra.env:ultra-v0",
+        agent_specs={AGENT_ID: spec},
+        scenario_info=scenario_info,
+        headless=headless,
+        timestep_sec=timestep_sec,
+        seed=seed,
+    )
+
+    agent = spec.build_agent()
+
+    for episode in episodes(num_episodes, etag=policy_class):
+        observations = env.reset()
+        state = observations[AGENT_ID]
+        dones, infos = {"__all__": False}, None
+        episode.reset()
+        experiment_dir = episode.experiment_dir
+
+        # save entire spec [ policy_params, reward_adapter, observation_adapter]
+        if not os.path.exists(f"{experiment_dir}/spec.pkl"):
+            if not os.path.exists(experiment_dir):
+                os.makedirs(experiment_dir)
+            with open(f"{experiment_dir}/spec.pkl", "wb") as spec_output:
+                dill.dump(spec, spec_output, pickle.HIGHEST_PROTOCOL)
+
+        while not dones["__all__"]:
+            if episode.get_itr(AGENT_ID) >= 1000000:  # 1M observation break
+                finished = True
+                break
+            evaluation_check(
+                agent=agent,
+                agent_id=AGENT_ID,
+                policy_class=policy_class,
+                episode=episode,
+                **eval_info,
+                **env.info,
+            )
+            action = agent.act(state, explore=True)
+            observations, rewards, dones, infos = env.step({AGENT_ID: action})
+            next_state = observations[AGENT_ID]
+
+            loss_output = agent.step(
+                state=state,
+                action=action,
+                reward=rewards[AGENT_ID],
+                next_state=next_state,
+                done=dones[AGENT_ID],
+            )
+            episode.record_step(
+                agent_id=AGENT_ID,
+                infos=infos,
+                rewards=rewards,
+                total_step=total_step,
+                loss_output=loss_output,
+            )
+            total_step += 1
+            state = next_state
+
+        episode.record_episode()
+        episode.record_tensorboard(agent_id=AGENT_ID)
+        if finished:
+            break
+
+    env.close()
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser("intersection-single-agent")
+    parser.add_argument(
+        "--task", help="Tasks available : [1, 2]", type=str, default="1"
+    )
+    parser.add_argument(
+        "--level",
+        help="Levels available : [easy, medium, hard, no-traffic]",
+        type=str,
+        default="easy",
+    )
```

Comment:
Same thing here obviously, should these be mandatory?

Reply:
They aren't mandatory since there are defaults. For training however, we might be able to think of a better api for users to select the env/task/scenario.

Reply:
Hmm.. yeah okay. We can maybe change this in a future ULTRA version.